src/emulator/functionsEmulator.ts

```diff
@@ -123,36 +132,30 @@ export class FunctionsEmulator implements EmulatorInstance {
       const reqBody = (req as RequestWithRawBody).rawBody;
       const proto = JSON.parse(reqBody.toString());
 
-      const runtime = FunctionsEmulator.startFunctionRuntime(
+      const worker = FunctionsEmulator.startFunctionRuntime(
         bundleTemplate,
         triggerId,
         EmulatedTriggerType.BACKGROUND,
         nodeBinary,
         proto
       );
 
-      runtime.events.on("log", (el: EmulatorLog) => {
+      worker.onLogs((el: EmulatorLog) => {
         if (el.level === "FATAL") {
           res.send(el.text);
         }
       });
 
-      // This "waiter" must be established before we block on "ready" since we expect
-      // this log entry to happen during the readying.
-      const triggerLogPromise = waitForLog(runtime.events, "SYSTEM", "triggers-parsed");
-
+      // TODO(samstern): Is this a race condition?  Could 'ready' happen before we're listening for it?
```

Comment:
Ok this sounds like a big (but good) change so I need your help unpacking it:

1) What is `BehaviorSubject`?
2) What do you mean "use the listener in the worker constructor to maintain it"
3) In fact, I may suggest to drop the waitForSystemLog and just expose promises for anything we want to listen to.  --> waitForSystemLog waits for the next log of a certain type so a promise isn't a suitable replacement

Reply:
Also since I see you approved: do you think these improvements should be made in this PR or should they be future improvements?

Reply:
First, the improvements can totally be done in a separate PR, and I approve this PR as-is. The log race condition issue is not the focus here and we can address that later.

I was suggesting `BehaviorSubject` from rxjs, but since we don't actually use Observables, I'd drop that idea since I don't want more things that we need to keep in mind when navigating the code base.

In details, I'd suggest that we keep the logic to wait on logs within `FunctionsRuntimeInstance`. In addition to `.exit`, it should also expose `.ready` and `.triggerParsed`, both promises. The promises themselves can of course be driven by logs, but we should not call `waitForLog` outside `FunctionsRuntimeInstance`. If we need more waits, we should expose each as a promise. In this way, we don't need to reason about the ordering outside.

Reply:
:+1: that makes sense, thanks for the explanation!  Will do those as a follow-up.